frontend/catalyst/pennylane_extensions.py

```diff
@@ -2633,22 +2635,38 @@ def callback_implementation(
     flat_args, in_tree = tree_flatten((args, kwargs))
     metadata = CallbackClosure(args, kwargs)
 
+    def to_shaped_array(ty):
+        if ty == inspect.Signature.empty:
+            return None
```

Comment:
Is the type annotation a reliable way to make this determination? I suppose we don't have too many other options since we can't trace the function 🤔 

The separation into dedicated callback types might allow us catch unintentional user errors in that case, like requiring a return type signature if the user uses the pure callback.

Reply:
> Is the type annotation a reliable way to make this determination? I suppose we don't have too many other options since we can't trace the function 🤔

I don't think there is another way except to rely on the user's annotation. That is also how JAX does it: https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html#exploring-jax-pure-callback

> The separation into dedicated callback types might allow us catch unintentional user errors in that case, like requiring a return type signature if the user uses the pure callback.

It is more explicit, you are correct. Another thing that this conversation inspired was: we should throw an exception if the return types mismatch (expected return type vs actual). That is not yet implemented in this PR. Will implement. Thanks!

Reply:
Oh nice, a runtime exception? That might be really necessary because if the types are different all kinds of problems could pop up in the rest of the compiled program.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/610/commits/936d63f86e964b73d80923383fe505f340dea32b